### PR TITLE
hashmaps: grow buckets when exceeding 75% load factor.

### PR DESCRIPTION
**Pull request by mrjbq7**
See issue #208.

### CODE REVIEW

File: lib-clay/hashmaps/hashmaps.clay [hunk 1]
```diff
@@ -84,6 +84,9 @@ overload put(a:HashMap[K,V], key:K, forward value:V) {
     ref bucket = a.buckets[i];
     bucket = UniqueValue(Node(key, value, move(bucket)));
     a.size += 1;
+    if (a.size > (size(a.buckets) * 0.75)) {
```
Comment (jckarter) [inline lib-clay/hashmaps/hashmaps.clay]: `* 3 / 4` would be better here; no point converting to float and back

File: lib-clay/hashmaps/hashmaps.clay [hunk 2]
```diff
@@ -99,9 +102,13 @@ overload index(a:HashMap[K,V], key:K) {
     if (not null?(ptr))
         return ref ptr^;
     ref bucket = a.buckets[i];
-    bucket = UniqueValue(Node(key, V(), move(bucket)));
+    var value = V();
+    bucket = UniqueValue(Node(key, value, move(bucket)));
     a.size += 1;
-    return ref bucket^.value;
+    if (a.size > (size(a.buckets) * 0.75)) {
```
Comment (jckarter) [inline lib-clay/hashmaps/hashmaps.clay]: same here

**Comment by mrjbq7** [comment]
Okay, how's that?

File: lib-clay/hashmaps/hashmaps.clay [hunk 3]
```diff
@@ -213,3 +220,28 @@ private findInBucket(_node:NodePtr[K,V], key:K) {
     }
     return null(V);
 }
+
+[K,V]
+private grow(a:HashMap[K,V]) {
+    var oldbuckets = a.buckets;
+
+    a.buckets <-- Vector[NodePtr[K,V]]();
```
Comment (jckarter) [inline lib-clay/hashmaps/hashmaps.clay:228]: This will leak memory. `var oldbuckets = a.buckets;` copies the old vector, then `a.buckets <-- Vector[...]();` overwrites `a.buckets` without releasing the memory. You can use `move` to safely move `a.buckets` into `oldbuckets`; after moving, `a.buckets` will be empty. Try this:

```
[K,V]
private grow(a:HashMap[K,V]) {
    var oldbuckets = move(a.buckets);

    a.size = 0;

    ...
}
```

File: lib-clay/hashmaps/hashmaps.clay [hunk 4]
```diff
@@ -213,3 +220,28 @@ private findInBucket(_node:NodePtr[K,V], key:K) {
...
+    a.size = 0;
+
+    try {
+        resize(a.buckets, size(oldbuckets) * 2);
+    } catch (ex) {
+        destroy(a.buckets);
+        throw ex;
+    }
```
Comment (jckarter) [inline lib-clay/hashmaps/hashmaps.clay]: Remove this `try` block; there's no need to manually destroy `a.buckets` if you use `move` like i described on line 228.

File: lib-clay/hashmaps/hashmaps.clay [hunk 5]
```diff
@@ -213,3 +220,28 @@ private findInBucket(_node:NodePtr[K,V], key:K) {
...
+
+    for(i in range(0, size(oldbuckets))) {
+        var node = &oldbuckets[i]^;
+        while (not null?(node)) {
+            put(a, node^.key, node^.value);
+            node = &node^.next^;
+        }
+    }
+
+    destroy(&oldbuckets);
```
Comment (jckarter) [inline lib-clay/hashmaps/hashmaps.clay]: You don't need to manually destroy local variables.

File: lib-clay/hashmaps/hashmaps.clay (hunk 4 above)
Comment (mrjbq7) [inline lib-clay/hashmaps/hashmaps.clay]: I'm trying to grow the a.buckets here to be twice as large -- using the convention from line 28 in `hashmaps.clay` where it tries to resize and destroys if it fails?  Is that not correct?

File: lib-clay/hashmaps/hashmaps.clay (hunk 5 above)
Comment (mrjbq7) [inline lib-clay/hashmaps/hashmaps.clay]: How do I know if my program leaks?  Is there valgrind support?

Comment (jckarter) [inline lib-clay/hashmaps/hashmaps.clay]: Yes. Run a clay program under valgrind and it should just work.

File: lib-clay/hashmaps/hashmaps.clay (hunk 4 above)
Comment (jckarter) [inline lib-clay/hashmaps/hashmaps.clay]: Line 28 is in the constructor and uses named return values. Because `returned` isn't a fully initialized value until the function returns, any partially initialized components must be manually destroyed if the constructor fails. If you don't use named return values or raw allocation you never need to use `<--` or manually call `destroy`.

Comment (mrjbq7) [inline lib-clay/hashmaps/hashmaps.clay]: So I can just `resize(a.buckets, size(oldbuckets) * 2)` and ignore any exception?

Comment (jckarter) [inline lib-clay/hashmaps/hashmaps.clay]: Yeah. If resize throws then the hashtable will be hosed, but all memory will at least be cleaned up.

`grow` could maybe be made more exception-robust if you initialized `newbuckets` outside of the object then did `swap(a.buckets, newbuckets);` at the end.

**Comment by mrjbq7** [comment]
Okay, updated the pull request with your suggestions...

File: lib-clay/hashmaps/hashmaps.clay [hunk 6]
```diff
@@ -99,9 +102,13 @@ overload index(a:HashMap[K,V], key:K) {
...
     ref bucket = a.buckets[i];
-    bucket = UniqueValue(Node(key, V(), move(bucket)));
+    var value = V();
+    bucket = UniqueValue(Node(key, value, move(bucket)));
     a.size += 1;
-    return ref bucket^.value;
+    if (a.size > (size(a.buckets) * 3 / 4)) {
+        grow(a);
+    }
+    return ref value;
```
Comment (jckarter) [inline lib-clay/hashmaps/hashmaps.clay]: This doesn't look right. You want to return a reference to the value inside the bucket, not the local variable. You should revert these changes other than the growth check.

Comment (mrjbq7) [inline lib-clay/hashmaps/hashmaps.clay]: Does bucket still point to the old bucket in the old list after grow() ? I wasn't sure.

Comment (jckarter) [inline lib-clay/hashmaps/hashmaps.clay]: No, since you allocate a new buckets vector, the `bucket` ref will be invalid. You'll need to `return ref a.buckets[i]^.value;` instead.

Comment (mrjbq7) [inline lib-clay/hashmaps/hashmaps.clay]: but `a.buckets[i]` is wrong since the index might have changed after rehash?  I was trying to hold onto the old location in memory or cache the old value somehow...

Comment (jckarter) [inline lib-clay/hashmaps/hashmaps.clay]: If the index changes, then you'll need to recalculate it. You can use the saved `bucket` reference in the case when it doesn't grow.

File: lib-clay/hashmaps/hashmaps.clay [hunk 7]
```diff
@@ -213,3 +220,21 @@ private findInBucket(_node:NodePtr[K,V], key:K) {
...
+
+    a.buckets <-- Vector[NodePtr[K,V]]();
+    a.size = 0;
+
+    resize(a.buckets, size(oldbuckets) * 2);
+
+    for(i in range(0, size(oldbuckets))) {
+        var node = &oldbuckets[i]^;
+        while (not null?(node)) {
+            put(a, node^.key, node^.value);
```
Comment (jckarter) [inline lib-clay/hashmaps/hashmaps.clay]: Since you're transferring values from the old bucket, you could move the key and value with `put(a, move(node^.key), move(node^.value));` here.

File: lib-clay/hashmaps/hashmaps.clay (hunk 6 above)
Comment (jckarter) [inline lib-clay/hashmaps/hashmaps.clay]: It looks like you could refactor `index` so that it `grow`s first if necessary then proceeds as it does currently. Does that seem reasonable? Then you wouldn't have any redundant calculation of the new key's bucket.

**Comment by mrjbq7** [comment]
Okay, incorporated that feedback.
